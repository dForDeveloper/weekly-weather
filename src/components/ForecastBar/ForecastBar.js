import React, { Component } from 'react';
import PropTypes from 'prop-types';

export class ForecastBar extends Component {
  constructor() {
    super();
    this.state = { isExpanded: false };
  }

  handleClick = () => {
    this.setState({ isExpanded: !this.state.isExpanded });
  }

  render() {
    const { isExpanded } = this.state;
    const {
      min,
      max,
      day,
      summary,
      icon,
      precipType,
      precipProbability,
      low,
      high,
      sunrise,
      sunset,
      wind,
      humidity
    } = this.props;
    const barStyle = {
      width: 100 * (high - low) / (max - min) + '%',
      marginLeft: 100 * (low - min) / (max - min) + '%'
    }
    return (
      <div className="ForecastBar">
        <div className={'ForecastBar--bar-container'}>
          <span className={icon +' small-icon'}></span>
          <span className='day'>{day}</span>
          <span className='bar-container'>
            <span className='bar' style={barStyle}>
              <span className='low'>{low}째</span>
              <span className='high'>{high}째</span>
            </span>
          </span>
          <button className='ForecastBar--button' onClick={this.handleClick}>
            {isExpanded ? '--' : '+'}
          </button>
        </div>
        {isExpanded &&
          <article className="article article--expanded">
            <header className='article--header'>
              <h2 className="article--h2"><div className={icon}></div>{summary}</h2>
              <span>Chance of {precipType}: {precipProbability}</span>
            </header>
            <div className="article--div">
              <p className="p">Low: {low}째</p>
              <p className="p">Sunrise: {sunrise}</p>
              <p className="p">Wind: {wind}</p>
              <p className="p">High: {high}째</p>
              <p className="p">Sunset: {sunset}</p>
              <p className="p">Humidity: {humidity}</p>
            </div>
          </article>
        }
      </div>
    );
  }
}

ForecastBar.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  day: PropTypes.string,
  summary: PropTypes.string,
  icon: PropTypes.string,
  precipType: PropTypes.string,
  precipProbability: PropTypes.string,
  low: PropTypes.number,
  high: PropTypes.number,
  sunrise: PropTypes.string,
  sunset: PropTypes.string,
  wind: PropTypes.string,
  humidity: PropTypes.string
}