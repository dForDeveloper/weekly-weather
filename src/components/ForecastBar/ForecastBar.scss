.ForecastBar {
  margin-bottom: 16px;
  box-shadow: $light-shadow;
  border-radius: 36px;
}

.ForecastBar--bar-container {
  width: 100%;
  height: 64px;
  display: grid;
  align-items: center;
  grid-template-columns: repeat(3, 48px) 1fr repeat(2, 48px);
  font-size: 20px;
  vertical-align: middle;
  padding: 8px 16px;
}

.small-icon {
  width: 48px;
  height: 48px;
}

.bar-container {
  height: 20px;
  margin-bottom: 6px;
  grid-column: 4/5;
}

.bar {
  height: 17px;
  background-color: #333333;
  border-radius: 10px;
  display: inline-block;
  vertical-align: middle;
  position: relative;
}

.high {
  vertical-align: middle;
  position: absolute;
  right: -42px;
  top: -3px;
}

.ForecastBar--button {
  border-radius: 24px;
  background-color: $blue;
  grid-column: 6/7;
  height: 36px;
  outline: none;
  border: none;
  color: #fff;
}

.low {
  position: absolute;
  left: -35px;
  top: -3px;
}

.day {
  margin-left: 8px;
}

.article--expanded {
  width: 100%;
  padding-bottom: 16px;
}