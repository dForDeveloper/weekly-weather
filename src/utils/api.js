export const fetchData = async (url) => {
  const response = await fetch(url);
  if (!response.ok) {
    const message = await response.json();
    throw Error(message);
  }
  return await response;
}