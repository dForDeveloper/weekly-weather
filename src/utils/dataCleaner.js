export const cleanData = (weather) => {
  const { temperature, apparentTemperature, summary } = weather.currently;
  const { minTemp, maxTemp } = weather.daily.data.reduce((acc, day) => {
    acc.minTemp = Math.round(Math.min(acc.minTemp, day.temperatureLow));
    acc.maxTemp = Math.round(Math.max(acc.maxTemp, day.temperatureHigh));
    return acc;
  }, { minTemp: 200, maxTemp: -200});
  const week = getWeekForecast(weather.daily.data);
  const firstDay = week.shift();
  const today = {
    ...firstDay,
    summary,
    temperature: Math.round(temperature),
    feelsLike: Math.round(apparentTemperature)
  };
  return { today, week, minTemp, maxTemp };
}

export const getWeekForecast = (data) => {
  return data.map(forecast => {
    const {
      summary,
      icon,
      precipType,
      precipProbability,
      time,
      temperatureLow,
      temperatureHigh,
      sunriseTime,
      sunsetTime,
      windSpeed,
      windBearing,
      humidity
    } = forecast;
    return {
      summary,
      icon,
      precipType,
      precipProbability: Math.round(precipProbability * 100) + '%',
      day: (new Date(time * 1000).toDateString().slice(0,3)),
      low: Math.round(temperatureLow),
      high: Math.round(temperatureHigh),
      sunrise: (new Date(sunriseTime * 1000))
        .toLocaleTimeString('en-US').replace(/:\d+/, ''),
      sunset: (new Date(sunsetTime * 1000))
        .toLocaleTimeString('en-US').replace(/:\d+/, ''),
      wind: Math.round(windSpeed) + ' mph ' + getWindDirection(windBearing),
      humidity: Math.round(humidity * 100) + '%'
    }
  });
}

export const getWindDirection = (windBearing) => {
  switch (Math.round(windBearing / 22.5) % 16) {
    case 0: return 'N';
    case 1: return 'NNE';
    case 2: return 'NE';
    case 3: return 'ENE';
    case 4: return 'E';
    case 5: return 'ESE';
    case 6: return 'SE';
    case 7: return 'SSE';
    case 8: return 'S';
    case 9: return 'SSW';
    case 10: return 'SW';
    case 11: return 'WSW';
    case 12: return 'W';
    case 13: return 'WNW';
    case 14: return 'NW';
    case 15: return 'NNW';
    default: return undefined;
  }
}