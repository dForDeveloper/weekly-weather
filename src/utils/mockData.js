export const mockHours = [
  {
    time: 1551132776,
    temperature: 55.55
  }
];

export const expectedHours = [
  {
    id: '',
    data: [{ x: '3:56 PM', y: 56 }]
  }
]

export const expectedWeek = [
  {
    summary: 'Snow (< 1 in.) until afternoon.',
    icon: 'snow',
    precipType: 'snow',
    precipProbability: '51%',
    day: 'Sat',
    low: 19,
    high: 37,
    sunrise: '6:56 AM',
    sunset: '5:17 PM',
    wind: '3 mph WNW',
    humidity: '66%'
  },
  {
    day: 'Sun',
    high: 42,
    humidity: '51%',
    icon: 'partly-cloudy-day',
    low: 23,
    precipProbability: '4%',
    precipType: 'snow',
    summary: 'Partly cloudy until afternoon.',
    sunrise: '6:31 AM',
    sunset: '5:24 PM',
    wind: '5 mph SSE'
  }
];

export const mockWeek = [
  {
    time: 1550905200,
    summary: 'Snow (< 1 in.) until afternoon.',
    icon: 'snow',
    sunriseTime: 1550929376,
    sunsetTime: 1550969177,
    moonPhase: 0.65,
    precipIntensity: 0.0038,
    precipIntensityMax: 0.029,
    precipIntensityMaxTime: 1550948400,
    precipProbability: 0.51,
    precipAccumulation: 0.63,
    precipType: 'snow',
    temperatureHigh: 36.58,
    temperatureHighTime: 1550959200,
    temperatureLow: 18.64,
    temperatureLowTime: 1551016800,
    apparentTemperatureHigh: 31.6,
    apparentTemperatureHighTime: 1550944800,
    apparentTemperatureLow: 10.43,
    apparentTemperatureLowTime: 1551016800,
    dewPoint: 16.25,
    humidity: 0.66,
    pressure: 1015.14,
    windSpeed: 2.78,
    windGust: 25.97,
    windGustTime: 1550962800,
    windBearing: 288,
    cloudCover: 0.59,
    uvIndex: 4,
    uvIndexTime: 1550952000,
    visibility: 7.53,
    ozone: 336.11,
    temperatureMin: 20.92,
    temperatureMinTime: 1550934000,
    temperatureMax: 36.58,
    temperatureMaxTime: 1550959200,
    apparentTemperatureMin: 14.85,
    apparentTemperatureMinTime: 1550937600,
    apparentTemperatureMax: 31.6,
    apparentTemperatureMaxTime: 1550944800
  },
  {
    time: 1550991600,
    summary: 'Partly cloudy until afternoon.',
    icon: 'partly-cloudy-day',
    sunriseTime: 1551015691,
    sunsetTime: 1551055644,
    moonPhase: 0.69,
    precipIntensity: 0.0005,
    precipIntensityMax: 0.0019,
    precipIntensityMaxTime: 1551009600,
    precipProbability: 0.04,
    precipAccumulation: 0.146,
    precipType: 'snow',
    temperatureHigh: 42.32,
    temperatureHighTime: 1551042000,
    temperatureLow: 22.97,
    temperatureLowTime: 1551092400,
    apparentTemperatureHigh: 37.9,
    apparentTemperatureHighTime: 1551045600,
    apparentTemperatureLow: 15.42,
    apparentTemperatureLowTime: 1551092400,
    dewPoint: 13.54,
    humidity: 0.51,
    pressure: 1022.33,
    windSpeed: 4.73,
    windGust: 11.93,
    windGustTime: 1551042000,
    windBearing: 163,
    cloudCover: 0.19,
    uvIndex: 4,
    uvIndexTime: 1551034800,
    visibility: 10,
    ozone: 314.42,
    temperatureMin: 18.64,
    temperatureMinTime: 1551016800,
    temperatureMax: 42.32,
    temperatureMaxTime: 1551042000,
    apparentTemperatureMin: 10.43,
    apparentTemperatureMinTime: 1551016800,
    apparentTemperatureMax: 37.9,
    apparentTemperatureMaxTime: 1551045600
  }
];

export const mockWeather = {
  currently: {
    time: 1550976078,
    summary: 'Partly Cloudy',
    icon: 'partly-cloudy-night',
    nearestStormDistance: 14,
    nearestStormBearing: 309,
    precipIntensity: 0,
    precipProbability: 0,
    temperature: 30.54,
    apparentTemperature: 21.27,
    dewPoint: 8.64,
    humidity: 0.39,
    pressure: 1021.54,
    windSpeed: 11.28,
    windGust: 24.75,
    windBearing: 267,
    cloudCover: 0.45,
    uvIndex: 0,
    visibility: 9.98,
    ozone: 335.67
  },
  daily: { data: mockWeek },
  hourly: {
    data: [
      {
        time: 1550973600,
        temperature: 37
      },
      {
        time: 1550977200,
        temperature: 33
      },
      {
        time: 1550980800,
        temperature: 31
      },
      {
        time: 1550984400,
        temperature: 29
      }
    ]
  }
}

export const expectedWeather = {
  graphData: [
    {
      id: '',
      data: [
        {
          x: '7:00 PM',
          y: 37,
        },
        {
          x: '8:00 PM',
          y: 33,
        },
        {
          x: '9:00 PM',
          y: 31,
        },
        {
          x: '10:00 PM',
          y: 29,
        },
      ]
    },
  ],
  maxTemp: 42,
  minTemp: 19,
  today: {
    day: 'Sat',
    feelsLike: 21,
    high: 37,
    humidity: '66%',
    icon: 'snow',
    low: 19,
    precipProbability: '51%',
    precipType: 'snow',
    summary: 'Partly Cloudy',
    sunrise: '6:56 AM',
    sunset: '5:17 PM',
    temperature: 31,
    wind: '3 mph WNW',
  },
  week:  [
    {
      day: 'Sun',
      high: 42,
      humidity: '51%',
      icon: 'partly-cloudy-day',
      low: 23,
      precipProbability: '4%',
      precipType: 'snow',
      summary: 'Partly cloudy until afternoon.',
      sunrise: '6:31 AM',
      sunset: '5:24 PM',
      wind: '5 mph SSE',
    },
  ],
}