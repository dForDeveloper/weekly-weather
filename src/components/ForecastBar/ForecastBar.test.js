import React from 'react';
import { shallow } from 'enzyme';
import { ForecastBar } from './ForecastBar';
import { expectedWeek } from '../../utils/mockData';

describe('ForecastBar', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <ForecastBar {...expectedWeek[0]} min={19} max={42} />
    );
  });

  it('should match the snapshow when is expanded is false', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  it('should match the snapshow when is expanded is true', () => {
    wrapper.setState({ isExpanded: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default state', () => {
    expect(wrapper.state('isExpanded')).toEqual(false);
  });

  it('should toggle isExpanded when the button is clicked', () => {
    expect(wrapper.state('isExpanded')).toEqual(false);
    wrapper.find('.ForecastBar--button').simulate('click');
    expect(wrapper.state('isExpanded')).toEqual(true);
  });
});