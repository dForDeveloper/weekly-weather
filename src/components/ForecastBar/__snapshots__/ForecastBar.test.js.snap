// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastBar should match the snapshow when is expanded is false 1`] = `
<div
  className="ForecastBar"
>
  <div
    className="ForecastBar--bar-container"
  >
    <span
      className="snow small-icon"
    />
    <span
      className="day"
    >
      Sat
    </span>
    <span
      className="bar-container"
    >
      <span
        className="bar"
        style={
          Object {
            "marginLeft": "0%",
            "width": "78.26086956521739%",
          }
        }
      >
        <span
          className="low"
        >
          19
          °
        </span>
        <span
          className="high"
        >
          37
          °
        </span>
      </span>
    </span>
    <button
      className="ForecastBar--button"
      onClick={[Function]}
    >
      +
    </button>
  </div>
</div>
`;

exports[`ForecastBar should match the snapshow when is expanded is true 1`] = `
<div
  className="ForecastBar"
>
  <div
    className="ForecastBar--bar-container"
  >
    <span
      className="snow small-icon"
    />
    <span
      className="day"
    >
      Sat
    </span>
    <span
      className="bar-container"
    >
      <span
        className="bar"
        style={
          Object {
            "marginLeft": "0%",
            "width": "78.26086956521739%",
          }
        }
      >
        <span
          className="low"
        >
          19
          °
        </span>
        <span
          className="high"
        >
          37
          °
        </span>
      </span>
    </span>
    <button
      className="ForecastBar--button"
      onClick={[Function]}
    >
      --
    </button>
  </div>
  <article
    className="article article--expanded"
  >
    <header
      className="article--header"
    >
      <h2
        className="article--h2"
      >
        <div
          className="snow"
        />
        Snow (&lt; 1 in.) until afternoon.
      </h2>
      <span>
        Chance of 
        snow
        : 
        51%
      </span>
    </header>
    <div
      className="article--div"
    >
      <p
        className="p"
      >
        Low: 
        19
        °
      </p>
      <p
        className="p"
      >
        Sunrise: 
        6:56 AM
      </p>
      <p
        className="p"
      >
        Wind: 
        3 mph WNW
      </p>
      <p
        className="p"
      >
        High: 
        37
        °
      </p>
      <p
        className="p"
      >
        Sunset: 
        5:17 PM
      </p>
      <p
        className="p"
      >
        Humidity: 
        66%
      </p>
    </div>
  </article>
</div>
`;
